### PR TITLE
Use bit-wise and operation to calculate sampling (#143)

### PR DESCRIPTION
**Pull request by gliptak**
Before opening your pull request, please make sure that you've:

- [x] [signed Uber's Contributor License Agreement](https://docs.google.com/a/uber.com/forms/d/1pAwS_-dA1KhPlfxzYLBqK6rsSWwRwH95OCCZrcsY5rk/viewform);
- [x] added tests to cover your changes;
- [x] run the test suite locally (`make test`); and finally,
- [x] run the linters locally (`make lint`).

Thanks for your contribution!

https://github.com/uber-go/zap/issues/143

### CODE REVIEW

File: zwrap/sample.go [hunk 1]
```diff
@@ -64,36 +64,44 @@ func (c *counters) Reset(key string) {
 // Sample returns a sampling logger. The logger maintains a separate bucket
 // for each message (e.g., "foo" in logger.Warn("foo")). In each tick, the
 // sampler will emit the first N logs in each bucket and every Mth log
-// therafter. Sampling loggers are safe for concurrent use.
+// thereafter (rounded down to nearest power of 2). Sampling loggers are safe for concurrent use.
```
Comment (akshayjshah) [inline zwrap/sample.go]: Thanks for fixing this misspelling :)

File: zwrap/sample.go [hunk 2]
```diff
@@ -64,36 +64,44 @@ func (c *counters) Reset(key string) {
...
-		Logger:     zl,
-		tick:       tick,
-		counts:     &counters{counts: make(map[string]*atomic.Uint64)},
-		first:      uint64(first),
-		thereafter: uint64(thereafter),
+		Logger: zl,
+		tick:   tick,
+		counts: &counters{counts: make(map[string]*atomic.Uint64)},
+		first:  uint64(first),
+		power:  uint64(power - 1),
```
Comment (akshayjshah) [inline zwrap/sample.go]: If `thereafter` isn't a power of two, let's also log an informative error using the passed-in logger. The message should include the desired value and the value we've rounded down to.

That way users get some feedback that they're misconfiguring their logger.

File: zwrap/sample_test.go [hunk 3]
```diff
@@ -97,7 +97,7 @@ func TestSampler(t *testing.T) {
 		for i := 1; i < 10; i++ {
 			tt.logFunc(sampler, i)
 		}
-		expected := buildExpectation(tt.level, 1, 2, 5, 8)
+		expected := buildExpectation(tt.level, 1, 2, 4, 6, 8)
```
Comment (akshayjshah) [inline zwrap/sample_test.go]: For completeness, let's also add a test that passes an exact power-of-two value for `thereafter` (which ensures that we're handling the rounding-down logic properly).

**Comment by akshayjshah** [comment]
One more important point: can you also add a benchmark for this? I'd like to compare against the master branch and make sure that the performance improvement is worth the somewhat surprising behavior.

**Comment by gliptak** [comment]
@akshayjshah Which benchmark method do you recommend to clone?
https://github.com/uber-go/zap/tree/master/benchmarks
Also how/where do you run the benchmarks?

File: zwrap/sample_test.go (hunk 3 above)
Comment (gliptak) [inline zwrap/sample_test.go]: added

File: zwrap/sample.go (hunk 2 above)
Comment (gliptak) [inline zwrap/sample.go]: using zl or onto stderr?

Comment (akshayjshah) [inline zwrap/sample.go]: Using `zl` - I'd like to make sure that the message ends up in the user's application logs.

**Comment by gliptak** [comment]
Master (modified to ```1<<13```):

```
BenchmarkZapSampleWithoutFields-4             	20000000	        77.4 ns/op	       0 B/op	       0 allocs/op
```

My branch:
```
BenchmarkZapSampleWithoutFieldsPowerOfTwo-4   	20000000	        75.2 ns/op	       0 B/op	       0 allocs/op
```

**Comment by gliptak** [comment]
@akshayjshah Are you good with the benchmark?

**Comment by akshayjshah** [comment]
@gliptak Thanks for adding the benchmark! Unfortunately, I don't think it's worth introducing this API limitation (sampling only powers of two) for a 2ns improvement.

We do very much appreciate your contribution, though - if you're interested in adding more to zap, I can tag you on some proposals that are less speculative.

**Comment by gliptak** [comment]
@akshayjshah Sure, please point me to other proposals.

**Pull request closed without merging**